### PR TITLE
Improved peer smart selector

### PR DESCRIPTION
**Pull request by devintegral**
- node.SmartPeerSelector.Next() optimization from O(n*n) to O(n log(n));
- tests of node.SmartPeerSelector;

### CODE REVIEW

File: src/node/peer_selector2_test.go [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+)
+
+const (
+	fakePeersCount = 50
+)
+
+func TestSmartSelector_Empty(t *testing.T) {
```
Comment (dzeckelev) [inline src/node/peer_selector2_test.go]: Do not use snake case please.

Comment (devintegral) [reply src/node/peer_selector2_test.go]: OK :(

**Comment by Maxime2** [comment]
Premature optimisation is the root of all evil.
First of all the algorithm of Smart peer selector is not finalised.
Secondly, the main problem with Peers and Peer selector is that they are not thread safe. This issue is not addressed in this pull-request. Thread safety was addressed in pull-request #101, but this PR made #101 unmergeable. 
Thirdly, speed improvements introduced in this pull-request weren't done for Random Peer Selector (the job half-done). 

@andrecronje why has this pull-request been made and then merged without any review from smart peer selector author?

**Comment by devintegral** [comment]
@Maxime2, let me answer:
1) Yeh, I've seen it at #120. But #114's changes still valuable;
2) It was SamuelMarks's decision to close #101 instead of conflicts resolving;
3) What's the problem? RandomPeerSelector's and SmartPeerSelector's improvements may be in the separately PRs. And RandomPeerSelector has no O(n*n) problem;

It was not optimisation for the sake of optimisation. It was commit of valuable result of the issue #93's research.

**Comment by Maxime2** [comment]
@devintegral This optimisation is valid, no question. But it doesn't solve the issue #93 and broke PR solving more severe problem of thread safety (in my estimate of severeness).

So the question is not about content of this PR itself, but rather about procedure how PRs are made and merged.

BTW, Smart peer selector sill has O(n^2) worst case estimate.

**Pull request merged**